scripts/bert/finetune_classifier.py

```diff
@@ -424,15 +431,61 @@ def train(metric):
         model.save_parameters(params_saved)
         logging.info('params saved in : {0}'.format(params_saved))
         toc = time.time()
-        logging.info('Time cost=%.1fs', toc - tic)
+        logging.info('Time cost=%.2fs', toc - tic)
         tic = toc
 
 
+def inference(metric):
+    """Inference function."""
+
+    logging.info('Now we are doing BERT classification inference on {} !'.format(ctx))
+    model = BERTClassifier(bert, dropout=0.1, num_classes=len(task.get_labels()))
+    model.hybridize()
+    model.load_parameters(model_parameters, ctx=ctx)
+
+    metric.reset()
+    step_loss = 0
+    tic = time.time()
+    for batch_id, seqs in enumerate(dev_data):
+        input_ids, valid_length, type_ids, label = seqs
+        out = model(input_ids.as_in_context(ctx),
+                    type_ids.as_in_context(ctx),
+                    valid_length.astype('float32').as_in_context(ctx))
+
+        ls = loss_function(out, label.as_in_context(ctx)).mean()
+
+        step_loss += ls.asscalar()
+        metric.update([label], [out])
+
+        if (batch_id + 1) % (args.log_interval) == 0:
+            metric_nm, metric_val = metric.get()
```

Comment:
Would you mind refactoring L461 - L469 into a function which can be used by both train() and inference()?

Reply:
Will do.

Reply:
ping for update

Reply:
There is different logging information in train() and inference() as epoch number and learning rate are not provided in inference(). Do you still think we can unify them into one `log_print` function?
```python
def log_print(batch_id, batch_num, metric, step_loss, log_interval, epoch_id=None, learning_rate=None):
    metric_nm, metric_val = metric.get()
    if not isinstance(metric_nm, list):
        metric_nm = [metric_nm]
        metric_val = [metric_val]

    # train log
    if (epoch_id is not None) and (learning_rate is not None):
        eval_str = '[Epoch %d Batch %d/%d] loss=%.4f, lr=%.7f, metrics: {' + \
                    ', '.join([i + ':%.4f' for i in metric_nm]) + '}'
        logging.info(eval_str, epoch_id + 1, batch_id + 1, batch_num, \
                             step_loss / log_interval, \
                             learning_rate, \
                             *metric_val)
    # inference log
    else:
        eval_str = '[Batch %d/%d] loss=%.4f, metrics: {' + \
                   ', '.join([i + ':%.4f' for i in metric_nm]) + '}'
        logging.info(eval_str, batch_id + 1, batch_num, \
                             step_loss / log_interval, \
                             *metric_val)
```

Reply:
I see. I didn't realize they are different. Nonetheless refactoring logging code as a function makes the training loop easier to read. What about adding `def log_train()` and `def log_eval()`?

Reply:
@eric-haibin-lin are you good with below changes? They're still looking redundant and confusing to me. But I'm not able to make them better. Any suggestion?
```python
def log_train(batch_id, batch_num, metric, step_loss, log_interval, epoch_id=None, learning_rate=None):
    metric_nm, metric_val = metric.get()
    if not isinstance(metric_nm, list):
        metric_nm = [metric_nm]
        metric_val = [metric_val]


    eval_str = '[Epoch %d Batch %d/%d] loss=%.4f, lr=%.7f, metrics: {' + \
                ','.join([i + ':%.4f' for i in metric_nm]) + '}'
    logging.info(eval_str, epoch_id + 1, batch_id + 1, batch_num, \
                         step_loss / log_interval, \
                         learning_rate, \
                         *metric_val)

def log_eval(batch_id, batch_num, metric, step_loss, log_interval):
    metric_nm, metric_val = metric.get()
    if not isinstance(metric_nm, list):
        metric_nm = [metric_nm]
        metric_val = [metric_val]
        
    eval_str = '[Batch %d/%d] loss=%.4f, metrics: {' + \
               ','.join([i + ':%.4f' for i in metric_nm]) + '}'
    logging.info(eval_str, batch_id + 1, batch_num, \
                         step_loss / log_interval, \
                         *metric_val)
```

Reply:
looks good

Reply:
Done. Thanks for your suggestion!